src/graphql/data/TokenPrice.ts

```diff
@@ -0,0 +1,173 @@
+    case TimePeriod.WEEK:
+      return 'WEEK'
+    case TimePeriod.MONTH:
+      return 'MONTH'
+    case TimePeriod.YEAR:
+      return 'YEAR'
+    case TimePeriod.ALL:
+      return 'MAX'
+  }
+}
+
+const allQuery = graphql`
+  query TokenPriceAllQuery($contract: ContractInput!) {
+    tokenProjects(contracts: [$contract]) {
+      name
+      markets(currencies: [USD]) {
+        priceHistory1H: priceHistory(duration: HOUR) {
+          timestamp
+          value
+        }
+        priceHistory1D: priceHistory(duration: DAY) {
+          timestamp
+          value
+        }
+        priceHistory1W: priceHistory(duration: WEEK) {
+          timestamp
+          value
+        }
+        priceHistory1M: priceHistory(duration: MONTH) {
+          timestamp
+          value
+        }
+        priceHistory1Y: priceHistory(duration: YEAR) {
+          timestamp
+          value
+        }
+        priceHistoryMAX: priceHistory(duration: MAX) {
+          timestamp
+          value
+        }
+      }
+      tokens {
+        chain
+        address
+        symbol
+        decimals
+      }
+    }
+  }
+`
+
+const query = graphql`
+  query TokenPriceSingleQuery($contract: ContractInput!, $duration: HistoryDuration!) {
+    tokenProjects(contracts: [$contract]) {
+      name
+      markets(currencies: [USD]) {
+        priceHistory: priceHistory(duration: $duration) {
+          timestamp
+          value
+        }
+      }
+      tokens {
+        chain
+        address
+        symbol
+        decimals
+      }
+    }
+  }
+`
+
+export function useTokenPriceQuery(address: string, chain: Chain, timePeriod: TimePeriod) {
+  const [prices, setPrices] = useState<PriceHistory>(TokenAPICache.checkPriceHistory(address, timePeriod))
+  const [error, setError] = useState<any>()
+  const [isLoading, setIsLoading] = useState(!prices)
+
+  /* To be called every time the data is successfully queried */
+  const updatePrices = useCallback(
+    (data: TokenPriceSingleQuery['response']) => {
+      const newPrices = data?.tokenProjects?.[0]?.markets?.[0]?.priceHistory
+      if (newPrices) {
+        TokenAPICache.setPriceHistory(newPrices, address, timePeriod)
+        setPrices(newPrices)
+      }
+    },
+    [timePeriod, address]
+  )
+
+  const fetchData = useCallback(() => {
+    fetchQuery<TokenPriceSingleQuery>(environment, query, {
+      contract: {
+        address,
+        chain,
+      },
+      duration: toHistoryDuration(timePeriod),
+    }).subscribe({
+      start: () => setIsLoading(true),
+      next: updatePrices,
+      error: setError,
+      complete: () => setIsLoading(false),
+    })
+  }, [setIsLoading, address, chain, timePeriod, updatePrices, setError])
+
+  useEffect(() => {
+    const cached = TokenAPICache.checkPriceHistory(address, timePeriod)
+    if (cached) {
+      setPrices(cached)
+    } else {
+      fetchData()
+    }
+  }, [address, timePeriod, fetchData])
+
+  return { error, isLoading, data: prices ?? [] }
```

Comment:
you'll return a new array for data every time - so just be mindful of object references. undefined might be better than an empty array